collect_app/src/main/java/org/odk/collect/android/formmanagement/previouslydownloaded/ServerFormsUpdateChecker.java

```diff
@@ -0,0 +1,79 @@
+package org.odk.collect.android.formmanagement.previouslydownloaded;
+
+import org.jetbrains.annotations.Nullable;
+import org.odk.collect.android.formmanagement.ServerFormDetails;
+import org.odk.collect.android.formmanagement.ServerFormsDetailsFetcher;
+import org.odk.collect.android.forms.FormRepository;
+import org.odk.collect.android.openrosa.api.FormApiException;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.Collections.emptyList;
+
+public class ServerFormsUpdateChecker {
+
+    private final ServerFormsDetailsFetcher serverFormsDetailsFetcher;
+    private final FormRepository formRepository;
+
+    public ServerFormsUpdateChecker(ServerFormsDetailsFetcher serverFormsDetailsFetcher, FormRepository formRepository) {
+        this.serverFormsDetailsFetcher = serverFormsDetailsFetcher;
+        this.formRepository = formRepository;
+    }
+
+    public List<ServerFormDetails> check() {
+        try {
+            List<ServerFormDetails> updatedForms = fetchUpdatedForms();
+            List<ServerFormDetails> newUpdates = new ArrayList<>();
+
+            for (ServerFormDetails serverFormDetails : updatedForms) {
+                String formHash = serverFormDetails.getHash();
+                String manifestFileHash = serverFormDetails.getManifestFileHash() != null ? serverFormDetails.getManifestFileHash() : "";
+
+                if (formRepository.getByLastDetectedUpdate(formHash, manifestFileHash) == null) {
+                    newUpdates.add(serverFormDetails);
+                    formRepository.setLastDetectedUpdated(serverFormDetails.getFormId(), formHash, manifestFileHash);
+                }
+            }
+
+            return newUpdates;
+        } catch (FormApiException e) {
+            return emptyList();
+        }
+    }
+
+    @Nullable
+    @SuppressWarnings("PMD.AvoidRethrowingException")
+    private List<ServerFormDetails> fetchUpdatedForms() throws FormApiException {
+        List<ServerFormDetails> formList = null;
+
+        try {
+            formList = serverFormsDetailsFetcher.fetchFormDetails();
+        } catch (FormApiException e) {
+            switch (e.getType()) {
+                case AUTH_REQUIRED:
+                    try {
+                        serverFormsDetailsFetcher.fetchFormDetails();
```

Comment:
What's the point of the single retry?

Reply:
This is the existing behaviour (see [here](https://github.com/getodk/collect/blob/8e6591552913f6a69894c6a8f2ccdefa2246fa0a/collect_app/src/main/java/org/odk/collect/android/tasks/ServerPollingJob.java#L93)). I'd actually meant to ask about this as it felt really weird. I'm thinking we can just get rid of it? Otherwise, we should write a test to keep it in (and hopefully describe what it's about).

Reply:
Wondered whether that might be old stuff. Hmm. In the foreground usage does could it be giving an opportunity to show the dialog or something?

Reply:
Yeah that'd make more sense. Probably want to view that as another issue/feature though. I think for the moment I'm happy for this to merge in like this if we add an issue around making the auto update prompt the user for creds if there is an auth error (like Manual and Match Exactly both will). How does that sound?

Reply:
Issue filed: https://github.com/getodk/collect/issues/3983